src/main/java/games/dominion/actions/Vassal.java

```diff
@@ -0,0 +1,127 @@
+package games.dominion.actions;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import core.AbstractGameState;
+import core.actions.AbstractAction;
+import core.actions.DoNothing;
+import core.components.Deck;
+import core.interfaces.IExtendedSequence;
+import games.dominion.DominionGameState;
+import games.dominion.DominionConstants.DeckType;
+import games.dominion.cards.CardType;
+import games.dominion.cards.DominionCard;
+
+/**
+ * Action to play the Vassal card. This card allows the player to draw the top card of their deck, and if it is an action card, they can choose to play it.
+ */
+public class Vassal extends DominionAction implements IExtendedSequence {
+
+    boolean executed = false;
+    CardType cardType;
+
+    public Vassal(int playerId) {
+        super(CardType.VASSAL, playerId);
+    }
+
+    public Vassal(int playerId, boolean dummy) {
+        super(CardType.VASSAL, playerId, dummy);
+    }
+
+    @Override
+    boolean _execute(DominionGameState state) {
+        Deck<DominionCard> drawPile = state.getDeck(DeckType.DRAW, player);
+        Deck<DominionCard> discardPile = state.getDeck(DeckType.DISCARD, player);
+
+        // Reshuffle if draw pile is empty
+        if (drawPile.getSize() == 0) {
+            discardPile.shuffle(state.getRnd());
+            drawPile.add(discardPile);
+            discardPile.clear();
+        }
+
```

Comment:
This is all done in DominionGameState.drawCard()...use that to draw cards to keep logic in one place